### PR TITLE
relabel: fix unmarshal

### PR DESCRIPTION
**Pull request by philchia**
Should not compare two regex ptr by !=, because they point to different variable

### CODE REVIEW

**Comment by roidelapluie** [comment]
Thank you for your contribution. We have looked at this during our bug scrub. It is possible to add a test to highlight the bug?

Also, for legal reasons, we require that all commits are signed with a [DCO](https://developercertificate.org/) before we can merge them.

**Comment by philchia** [comment]
Done

**Comment by philchia** [comment]
@roidelapluie PTAL

**Comment by roidelapluie** [comment]
In your test, you add it expliclity. Therefore it should probably fail, even if it is the default value.

File: model/relabel/relabel_test.go [hunk 1]
```diff
@@ -758,3 +763,41 @@ func BenchmarkRelabel(b *testing.B) {
...
+		name    string
+		fields  fields
+		wantErr bool
+	}{
+		{
+			name: "keepequal",
+			fields: fields{
+				Source: `
+action: keepequal
+regex: (.*)
```
Comment (roidelapluie) [inline model/relabel/relabel_test.go:782]: I expect this config to fail as the value is set explicitly.

Comment (philchia) [reply model/relabel/relabel_test.go:782]: You may got this value from a normal yaml marshal, i expect this not to fail

Comment (philchia) [reply model/relabel/relabel_test.go:782]: Otherwise, we need to implement MarshalYAML for Config

**Comment by bboreham** [comment]
In my opinion the original code is correct: it checks whether the value has been set, not whether it has been set different to the default.

**Comment by bboreham** [comment]
> Got this value from a normal yaml marshal, maybe we need to implement MarshalYAML for Config.

We discussed at the bug scrub, and agreed with this suggestion.  Implement MarshalYAML to remove the regexp field if it was unset (set to default).

This would also improve the `/config` page in the Prometheus UI, as the displayed rules would be more like the input.

**Comment by go-follow** [comment]
We are really looking forward to keepequal. Please tell me when you plan to merge this PR?

**Comment by bboreham** [comment]
There are no plans to merge this PR.  As you can see in the above conversation, the author suggested a different approach, and the Prometheus maintainers agreed with that suggestion.  The next step is for someone to implement that different approach.

**Comment by beorn7** [comment]
The different approach is implemented as #14004. We are still waiting to clarify one last concern over there, but for now, I assume #14004 will get merged. I'm closing this PR in lieu if #14004.

**Pull request closed without merging**